distributed/client.py

```diff
@@ -242,6 +243,22 @@ def _bind_late(self):
                     pass
                 else:
                     handler(key=self.key)
+        if self._client:
+            try:
+                assert self._state is self._client.futures[self._tkey]
+            except (AssertionError, KeyError):
+                raise RuntimeError(
+                    textwrap.dedent(
+                        """\
+                Critical error encountered during Future initialization. This
+                typically occurs when passing around Future objects without
+                handling the Client lifecycle explicitly. If you encounter this,
+                please ensure to initialize a Client object yourself before
```

Comment:
This seems like an improvement and I don't want to add much friction, but in case the perspective of a naive user is relevant:

If I'd written the code in the test below, and then I read "please ensure to initialize a Client object yourself before interacting with the Future object" then I think my reaction would be:

"But I did initialize a client object before interacting with the future! The very beginning of my test initializes a client, and then the client is explicitly right there in `client.submit` and `client.map` whenever I interact with the future."

I'm sure I'm wrong in some sense, and maybe I see a little bit of why but not sure I understand the actual situation well enough to suggest language that would have worked on me.

Reply:
Changed it to 

> Critical error encountered during Future initialization.
> 
> This typically occurs when interacting with Future objects
> directly inside of a Task without initializing a Client
> explicitly inside of this Task first.
> If you encounter this, please ensure to initialize a Client
> object yourself before interacting with the Future object
> directly. See also
> https://distributed.dask.org/en/stable/api.html#distributed.worker_client

Is this easier to understand for a non-expert?

Reply:
The link also shows an example so I am hoping this will clear things up even if users do not comprehend exactly the details of the exception message but I'm happy to take further suggestions about the concrete phrasing

Reply:
Yeah, I think the new text and the link are pretty clear about how to fix it.

(Still very confusing though! At this point as a user I'm thinking "wait what?! `client.submit(function, arg)` without making a client inside the function is super common across all the dask usage, docs, tutorials, etc. that I've seen. Why am I being told now that it's bad?"

Reply:
> without making a client inside the function is super common across all the dask usage, docs, tutorials, etc. that I've seen. Why am I being told now that it's bad?")

```python
f = client.submit(foo)
g = client.submit(bar, f)
```

is fine and very, very common. We'll resolve this for the user and the user will only interact with the result of `f`, not the actual `Future` object.

```python
f = client.submit(foo)
g = client.submit(bar, OpaqueWrapper(f))
```
will give the user the actual `Future` object and not the result of that `Future` object. This is where it gets tricky.

I don't think this difference can be explained in an exception message. It's also relatively rare

Reply:
Maybe a concrete code snippet in the error message would help make this clearer?

```
For example, instead of:

dask_future = client.submit(...)
def func():
    x = dask_future.result()
f = client.submit(func)

Try:

dask_future = client.submit(...)
def func():
    # Explicitly initialize client within task before retrieving future's result
    with distributed.worker_client() as c:
        x = dask_future.result()
f = client.submit(func)

```